### DISCUSSION TITLE
Feels too early for v2

### DISCUSSION THREAD

**Discussion by mattmast** [moonrepo/moon#1318] [General]
Our team is currently evaluating monorepo tools. I've been super excited about `moon` after discovering it. For the most part a PoC using `moon` has been easy to setup. We are a smaller team so we don't need anything crazy. I love that you are distinguishing yourselves as a repo management tool and would get value out of the other features.  

However, compared to `nx`, `moon` hasn't been around as long. It doesn't quite have the test of time. There's less community adoption so as a team we've been hesitant to commit. We were about to take the plunge but then saw the v2 announcement and it has us pulling back not knowing how big of change v2 will bring and if it will be an improvement.

I'm not a fan of `nx`, one reason is the plugin system. I don't want to learn another abstraction layer over all our tooling (i.e. typescript, eslint, jest, esbuild, webpack, etc.).  I want it as close to the configuration and the tooling as possible, and shell scripts get me there. 

I guess I wanted to share my perspective because I really like what your team is bringing to the table but have concerns over adoption regarding v2.

**Comment by milesj** [comment]
@mattmast Thanks for your concern, but let me clarify a few things around plugins that'll hopefully ease your mind. To start, moon plugins are nothing like Nx plugins. 

Nx plugins are more like wrappers around specific tools to automate tasks/configs, and are either authored by the Nx team themselves, or by the community. Nx kind of forces you to use plugins, which we agree, is not a nice abstraction.

moon plugins are more of an internal only thing, and are split into 2 categories right now: extensions and platforms. 

- An extension is simply a plugin for running one-off functionality, and is not required by consumers to know, use, or write, but is available if they need it. https://moonrepo.dev/docs/guides/extensions
- The platform system hasn't landed yet and is primarily what v2 will be about. A platform provides deep language integration (node.js, rust, go, etc) into moon, so that those languages better interop with moon's internals and pipeline. This includes stuff like lockfile parsing, manifesting reading, and more.

Both of these plugins should be entirely transparent to consumers and will just automatically work behind the scenes. Furthermore, consumers don't really need to know about them, or write them themselves, they kind of just exist. However, we call them plugins because if a consumer _does want_ to build and use their own plugins, they can! For example, say they want to add C++ support and we don't provide it yet, they can fill that gap themselves.

As for your other concerns, v2 is all about just moving to a plugin system that allows us to build these plugins in isolation, outside of moon's Rust core. All the other features of moon, like configuration, tasks, projects, the pipeline, etc, will not be changing.

**Reply by mattmast** [reply]
Thank you so much for the quick and detailed response!  That does clarify a lot. I'm sorry I read too much into it initially. 

I guess I saw "big plugin re-architecture", v2 as being a major semantic version, and feared there would be lots of breaking changes to configs or to the CLI for core functionality.  

I'm glad to hear we agree on nx. :)

**Reply by milesj** [reply]
We do have some breaking changes for v2, but it's mostly just renaming settings, or moving things around a bit. Should _hopefully_ be a simple upgrade!
